elastic/kibana
# added source filtering

# moved to #5981
- [x] fetch field if:
  - [x] explicitly selected in discover
  - [x] show in a table
  - [x] used in an aggregate

---

close #4366 

I am not sure if this PR is a complete solution for the issue, but I believe it is a part of it.

This PR adds a configuration on the index pattern level that allows to define which field to exclude using the [source filtering](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-source-filtering.html) API. There is a new tab in which you can set the JSON to specify what to exclude.
The fields that are actually retrieved are indicated on the `Fields` tab as an additional column: if it is retrieved, the row is checked.

![2016-01-12-104207_1920x1848_scrot](https://cloud.githubusercontent.com/assets/795879/12261682/9a2218a2-b91a-11e5-97be-5ae9ead6fb76.png)

To test it:
- Create a small index:

``` JSON
curl -XPUT 'http://localhost:9200/country/Country/1' -d '
{
  "name": "Italy",
  "description": "pasta"
}'
```
- Add the `country` index and go to the **Retrieved Fields** tab. Exclude the **description** field from being retrieved. Add the following object to do so:

``` json
{
 "exclude": "description"
}
```

![2016-01-12-105353_1920x1848_scrot](https://cloud.githubusercontent.com/assets/795879/12261796/6865e2f2-b91b-11e5-8fe1-626ca17bc651.png)
- The **description** is now unchecked, meaning it is not retrieved.

![2016-01-12-103741_1920x1848_scrot](https://cloud.githubusercontent.com/assets/795879/12261506/71bcc412-b919-11e5-9e1a-8335d58f2c74.png)
- Create a saved search for the countries and then add a search visualisation to a dashboard. You will see that the field **description** does not appear.

![2016-01-12-104756_1920x1848_scrot](https://cloud.githubusercontent.com/assets/795879/12261610/0da081b6-b91a-11e5-845c-ca829433f5c0.png)

---

What is not provided in this PR but could be necessary: a way to display an **excluded** field for a document. Do you see this mandatory, or is the proposed solution good enough ?
